### PR TITLE
search: Refactor suggestions and add integration tests for them

### LINKED ISSUE #18621: search: Add integration tests for search suggestions
We don't have any integration tests for search suggestions. This allowed #18619 to sneak through, which would've been caught with even just a sanity test.

### LINKED ISSUE #18622: search: Use proper typing for search suggestions
Since search suggestions store all the resolver types in a nameless `interface{}` type, and panic if an unsupported type is used, making a change to those types isn't validated by static analysis. We can add some tests to help with this, but ideally, we'd refactor a bit to make the search result resolver more statically typed. This contributed to the issues fixed by #18619

### PR DESCRIPTION
**Pull request by camdencheek**
Motivation: earlier today, a change I made that touched search suggestions caused a panic in frontend that took down cloud. I identified two things that would have helped prevent that outage. First was having integration tests that covered suggestions. Even a minimal test would have caught this panic. Second was refactoring the search suggestion code so that it relied less on `interface{}` so that static analysis would catch type errors. 

Fixes #18621 
Fixes #18622

Overview of the changes:
- Make `SearchSuggestionResolver` an interface
- Create wrapper type for each type of suggestion that implements `SearchSuggestionResolver`
- Move core suggestion types into `search_suggestions.go`
- Convert consumers of `newSearchSuggestionResolver` to use the wrapper types directly
- Add an integration test

I apologize in advance for the large changes -- I struggled to split this one up.

### CODE REVIEW

File: cmd/frontend/graphqlbackend/search_suggestions.go [hunk 1]
```diff
@@ -72,9 +213,9 @@ func (r *searchResolver) Suggestions(ctx context.Context, args *searchSuggestion
 		}
```
Comment (camdencheek) [inline cmd/frontend/graphqlbackend/search_suggestions.go:213]: Why do we only do suggestions for `type:file`?

Comment (rvantonder) [reply cmd/frontend/graphqlbackend/search_suggestions.go:213]: Honestly, don't know. The search suggestions code needs a lot of love. There's a PR somewhere that uncovers some of the history, but I can't find it right now. Note that `type:file` is a misnomer currently and actually means "search for content".

**Review by rvantonder — APPROVED** [review]
@camdencheek be like:




-> this should have been reviewed sooner, sorry for it going stale.

File: cmd/frontend/graphqlbackend/search_suggestions.go [hunk 2]
```diff
@@ -26,6 +25,148 @@ import (
...
+// baseSuggestionResolver implements all the To* methods, returning false for all of them.
+// Its intent is to be embedded into other suggestion resolvers to simplify implementing
+// searchSuggestionResolver.
+type baseSuggestionResolver struct{}
+
+func (baseSuggestionResolver) ToRepository() (*RepositoryResolver, bool) { return nil, false }
+func (baseSuggestionResolver) ToFile() (*GitTreeEntryResolver, bool)     { return nil, false }
+func (baseSuggestionResolver) ToGitBlob() (*GitTreeEntryResolver, bool)  { return nil, false }
+func (baseSuggestionResolver) ToGitTree() (*GitTreeEntryResolver, bool)  { return nil, false }
+func (baseSuggestionResolver) ToSymbol() (*symbolResolver, bool)         { return &symbolResolver{}, false }
```
Comment (keegancsmith) [inline cmd/frontend/graphqlbackend/search_suggestions.go:59]:
```suggestion
func (baseSuggestionResolver) ToSymbol() (*symbolResolver, bool)         { return nil, false }
```

File: cmd/frontend/graphqlbackend/search_suggestions.go [hunk 3]
```diff
@@ -26,6 +25,148 @@ import (
...
+	gitTreeEntry *GitTreeEntryResolver
+	score        int
+}
+
+func (g gitTreeSuggestionResolver) Score() int    { return g.score }
+func (g gitTreeSuggestionResolver) Length() int   { return len(g.gitTreeEntry.Path()) }
+func (g gitTreeSuggestionResolver) Label() string { return g.gitTreeEntry.Path() }
+func (g gitTreeSuggestionResolver) ToFile() (*GitTreeEntryResolver, bool) {
+	return g.gitTreeEntry, true
+}
```
Comment (keegancsmith) [inline cmd/frontend/graphqlbackend/search_suggestions.go:89]: I didn't realise GraphQL supported multiple `To...` returning true. I wonder what happens, I'm guessing everything just works :)

File: cmd/frontend/graphqlbackend/search_suggestions.go [hunk 4]
```diff
@@ -26,6 +25,148 @@ import (
...
+		// Sort by score
+		a, b := s[i], s[j]
+		if a.Score() != b.Score() {
+			return a.Score() > b.Score()
+		}
+		// Prefer shorter strings for the same match score
+		// E.g. prefer gorilla/mux over gorilla/muxy, Microsoft/vscode over g3ortega/vscode-crystal
+		if a.Length() != b.Length() {
+			return a.Length() < b.Length()
+		}
```
Comment (keegancsmith) [inline cmd/frontend/graphqlbackend/search_suggestions.go:154]: It looks like we can get rid of Length and just rely on length of Label?

**Review by keegancsmith — APPROVED** [review]

**Pull request merged**